### PR TITLE
Trinity 2.2.0

### PR DESCRIPTION
**Pull request by abretaud**
A new Trinity version + some other changes:
- a new abundance estimation method: salmon
- macros and tool suite to be more IUC-friendly
- one or two options that I needed to add
- 2 new tools for trinity scripts that I needed

I have added a tool dependency xml for this new 2.2.0 version, but I didn't create a new one for salmon, if anyone needs it tell me.

I *think* there may still be a failing test, but I'm not sure if it comes from my config or not, we'll see

### CODE REVIEW

**Comment by abretaud** [comment]
Ok, run_DE_analysis test failed, and I'm not sure what to do...
This is the log I get:
````
Fatal error: Exit code 25 (Error in Trinity execution)
Got 9 samples, and got: 10 data fields.
Header: wt_37_2	wt_37_3	wt_37_1	wt_GSNO_3	wt_GSNO_1	wt_GSNO_2	wt_ph8_3	wt_ph8_1	wt_ph8_2
Next: TR24|c0_g1_i1	90.00	67.00	85.00	36.00	35.00	34.00	222.00	196.00	201.00

$VAR1 = {
          'wt_ph8_3' => 7,
          'wt_GSNO_1' => 5,
          'wt_37_1' => 3,
          'wt_GSNO_3' => 4,
          'wt_ph8_2' => 9,
          'wt_37_3' => 2,
          'wt_37_2' => 1,
          'wt_ph8_1' => 8,
          'wt_GSNO_2' => 6
        };
Contrasts to perform are: $VAR1 = [
          [
            'wt_37',
            'wt_GSNO'
          ],
          [
            'wt_37',
            'wt_ph8'
          ],
          [
            'wt_GSNO',
            'wt_ph8'
          ]
        ];
Loading required package: S4Vectors
Loading required package: stats4
Loading required package: BiocGenerics
Loading required package: parallel

Attaching package: 'BiocGenerics'

The following objects are masked from 'package:parallel':

    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
    clusterExport, clusterMap, parApply, parCapply, parLapply,
    parLapplyLB, parRapply, parSapply, parSapplyLB

The following objects are masked from 'package:stats':

    IQR, mad, xtabs

The following objects are masked from 'package:base':

    Filter, Find, Map, Position, Reduce, anyDuplicated, append,
    as.data.frame, as.vector, cbind, colnames, do.call, duplicated,
    eval, evalq, get, grep, grepl, intersect, is.unsorted, lapply,
    mapply, match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    rank, rbind, rownames, sapply, setdiff, sort, table, tapply, union,
    unique, unlist, unsplit

Loading required package: IRanges
Loading required package: GenomicRanges
Loading required package: GenomeInfoDb
Loading required package: SummarizedExperiment
Error: This is R 3.1.2, package 'SummarizedExperiment' needs >= 3.2
In addition: Warning messages:
1: package 'DESeq2' was built under R version 3.2.2 
2: package 'S4Vectors' was built under R version 3.2.2 
3: package 'BiocGenerics' was built under R version 3.2.2 
4: In eval(quote({ : bytecode version mismatch; using eval
5: In .recacheSubclasses(def@className, def, doSubclasses, env) :
  undefined subclass "externalRefMethod" of class "expressionORfunction"; definition not updated
6: package 'IRanges' was built under R version 3.2.2 
7: In .recacheSubclasses(def@className, def, doSubclasses, env) :
  undefined subclass "externalRefMethod" of class "functionORNULL"; definition not updated
8: package 'GenomicRanges' was built under R version 3.2.2 
9: package 'GenomeInfoDb' was built under R version 3.2.2 
Execution halted
Error, cmd: R --vanilla -q < input.matrix.wt_37_vs_wt_GSNO.DESeq2.Rscript died with ret (256)  at /tmp/tmp29HRyD/job_working_directory/000/3/conda-env/bin/run_DE_analysis.pl line 712.
````
So it looks like a R version problem, there are 3 requirements:
```
        <requirement type="package" version="1.10.0">bioconductor-deseq2</requirement>
        <requirement type="package" version="3.14.0">bioconductor-edger</requirement>
        <requirement type="package" version="3.28.10">bioconductor-limma</requirement>
```
When installed independently, deseq2 depends on R 3.2.2, but both edger and limma depends on R 3.1.2.
I don't know if it's the expected behaviour, but it seems like it's the R binary from the last package which is kept. Does anyone know what to do? I would try to put deseq2 at the end, but I'm a little worried by the [warning](https://github.com/galaxyproject/tools-iuc/blob/master/tools/trinity/run_DE_analysis.xml#L5)...

**Comment by abretaud** [comment]
Right, I changed the requirements order and the tests are running fine now. It shouldn't have impact on tool_dependency method as deseq2 package name is different from conda and I didn't touch it.

Conda tests are all ok, but the build is still failing for some strange reason, maybe flake8 problem?
```
./tools/cwpair2/cwpair2_util.py:9:0: E402 module level import not at top of file
./tools/repmatch_gff3/repmatch_gff3_util.py:10:0: E402 module level import not at top of file
flake8.main.application   MainProcess   3218 INFO     Found a total of 2239 violations and reported 2
```

**Comment by nsoranzo** [comment]
@abretaud I've fixed the flake8 issues on `master` branch (these were due to the new flake8 3.0.0), and restarted your build, it's green now!

File: tools/trinity/.shed.yml [hunk 1]
```diff
@@ -11,3 +11,9 @@ name: trinity
 owner: iuc
```
Comment (nsoranzo) [inline tools/trinity/.shed.yml]: You can remove `description:` and `name:` at lines 4 and 10.

File: tools/trinity/align_and_estimate_abundance.xml [hunk 2]
```diff
@@ -1,18 +1,35 @@
...
-    </requirements>
-    <stdio>
-        <exit_code range="1:"/>
-    </stdio>
+        <requirement type="package" version="0.6.0">salmon</requirement>
+    </expand>
+    <expand macro="stdio"/>
     <command><![CDATA[
-        ln -s $transcripts input.fa
+        ln -s $transcripts input.fa &&
```
Comment (nsoranzo) [inline tools/trinity/align_and_estimate_abundance.xml]: s/$transcripts/'$transcripts'/

File: tools/trinity/align_and_estimate_abundance.xml [hunk 3]
```diff
@@ -1,18 +1,35 @@
...
+                ln -s $inputs.right_input paired_right.fa
+            #else:
+                ln -s $inputs.left_input paired_left.fq &&
+                ln -s $inputs.right_input paired_right.fq
+            #end if
+        #else:
+            #if $inputs.left_input.is_of_type('fasta'):
+                ln -s $inputs.input single.fa
+            #else:
+                ln -s $inputs.input single.fq
```
Comment (nsoranzo) [inline tools/trinity/align_and_estimate_abundance.xml]: Single-quote the various `$inputs....` in lines 20-30.

File: tools/trinity/filter_low_expr_transcripts.xml [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            #if str($isoform_filter.highest_iso_only) == "yes":
+                --highest_iso_only
+            #else if str($isoform_filter.min_pct_dom_iso):
+                --min_pct_dom_iso $isoform_filter.min_pct_dom_iso
+            #end if
+
+            > "$filtered"
+    ]]></command>
+    <inputs>
+        <param format="fasta" name="assembly" type="data" label="Trinity assembly"/>
```
Comment (nsoranzo) [inline tools/trinity/filter_low_expr_transcripts.xml]: Can you add `help="--transcripts"` and similarly for the other params?

File: tools/trinity/filter_low_expr_transcripts.xml [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+                </when>
+                <when value="no">
+                    <param format="tabular" name="gene_trans_map" type="data" label="Gene to transcript correspondence ('gene(tab)transcript' lines)" />
+                </when>
+            </conditional>
+
+        </section>
+    </inputs>
+    <outputs>
+        <data format="tabular" name="filtered" label="${tool.name} on ${on_string}: filtered low expression transcripts" from_work_dir="map.tsv"/>
```
Comment (nsoranzo) [inline tools/trinity/filter_low_expr_transcripts.xml]: Is `from_work_dir` necessary here? You are redirecting stdout to this output in `<command>`.

File: tools/trinity/filter_low_expr_transcripts.xml [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+            <param name="matrix" value="filtered/counts.matrix"/>
+            <param name="assembly" value="raw/Trinity.fasta" ftype="fasta"/>
+            <param name="isoform_filter|min_pct_dom_iso" value="20"/>
+            <output name="filtered" file="filtered/Trinity_filtered_pct_iso.fasta" />
+        </test>
+    </tests>
+    <help>
+        Trinity_ assembles transcript sequences from Illumina RNA-Seq data.
+        This tool produce a file containing correspondance between gene ids and transcript ids based on the name of transcripts assembled by Trinity.
+        The output file is intended to be used by the "Align reads and estimate abundance" tool.
```
Comment (nsoranzo) [inline tools/trinity/filter_low_expr_transcripts.xml]: Copy-paste from the help of `gene_to_trans_map.xml` , can you fix please?

File: tools/trinity/gene_to_trans_map.xml [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    <expand macro="requirements"/>
+    <expand macro="stdio"/>
+    <command><![CDATA[
+       get_Trinity_gene_to_trans_map.pl "$assembly" > "$map"
+    ]]></command>
+    <inputs>
+        <param format="fasta" name="assembly" type="data" label="Trinity assembly"/>
+    </inputs>
+    <outputs>
+        <data format="tabular" name="map" label="${tool.name} on ${on_string}: Genes to transcripts map" from_work_dir="map.tsv"/>
```
Comment (nsoranzo) [inline tools/trinity/gene_to_trans_map.xml]: Is `from_work_dir` necessary here? You are redirecting stdout to this output in `<command>`.

File: tools/trinity/trinity.xml [hunk 8]
```diff
@@ -127,8 +134,9 @@
                 </when>
             </conditional>
 
-
             <param format="fasta" name="long_reads" type="data" optional="true" label="Error-corrected or circular consensus (CCS) pac bio reads" help="Experimental feature! Long reads must be in the same orientation as short reads if they are strand specific"/>
+
+            <param name="min_kmer_cov" type="integer" optional="true" value="1" min="1" label="Minimum count for K-mers to be assembled"/>
```
Comment (nsoranzo) [inline tools/trinity/trinity.xml]: `help="--min_kmer_cov"`

**Comment by abretaud** [comment]
I added `argument="--xxx"` instead of `help="--xxx"`, is it ok?
The other corrections are made

**Comment by nsoranzo** [comment]
@abretaud Awesome! :+1:

**Comment by bgruening** [comment]
You guys rock!

**Comment by abretaud** [comment]
Thanks! I'm pushing to TS manually as automatic push failed

**Comment by lecorguille** [comment]
Too late! I was coming to add a <stdio> tag on Trinity 2.1.1.1
GG! @abretaud

**Comment by peterjc** [comment]
Looks like an issue from this on TravisCI, https://travis-ci.org/galaxyproject/tools-iuc/jobs/147738492

```
cd '/home/travis/build/galaxyproject/tools-iuc/tools/trinity' && git rev-parse HEAD
cd '/home/travis/build/galaxyproject/tools-iuc/tools/trinity' && git diff --quiet
Could not update abundance_estimates_to_matrix
Unexpected response from galaxy: 500: {"content_alert": "", "err_msg": "Unable to locate repository with name package_trinity_2_2_0 and owner iuc.    The tool_dependencies.xml file contains an invalid <repository> tag."}
Failed to update a repository.
```

**Comment by nsoranzo** [comment]
@peterjc See https://github.com/galaxyproject/tools-iuc/pull/894#issuecomment-235572900 and https://github.com/galaxyproject/tools-iuc/issues/898 .

**Pull request merged**